packages/odyssey-react-mui/src/test-selectors/querySelector.ts

```diff
@@ -11,188 +11,238 @@
+          keyof LocalFeatureTestSelector["selector"]["options"],
+          string | RegExp
+        >;
+      }
+    : object);
+
+export const querySelector =
+  <LocalFeatureTestSelector extends FeatureTestSelector>(
+    /**
+     * Selectors object including features and accessible text selections.
+     */
+    featureTestSelector: LocalFeatureTestSelector,
+  ) =>
+  <LocalQueryMethod extends QueryMethod = "get">(
+    props: {
+      /**
+       * Refers to Testing Library's canvas. This is usually `screen`, but Storybook uses `within(canvas)`.
+       */
+      element: HTMLElement;
+    } & InnerQuerySelectorProps<LocalFeatureTestSelector, LocalQueryMethod>,
+  ) => {
+    const { element: containerElement, queryMethod } = props;
+    const localQueryMethod = queryMethod || ("get" as const);
+    const querySelectorOptions = "options" in props ? props.options : undefined;
+    const role = "role" in props ? (props.role as AriaRole) : undefined;
+
+    // This `let` is difficult to make into a `const`. It makes the code unreadable.
+    let capturedElement: HTMLElement | null = null;
+
+    if ("selector" in featureTestSelector && querySelectorOptions) {
+      const sharedProps = {
+        element: containerElement,
+        queryMethod: localQueryMethod,
+        queryOptions: Object.fromEntries(
+          Object.entries(featureTestSelector.selector.options).map(
+            ([testSelectorsKey, testingLibraryKey]) => [
+              testingLibraryKey,
+              querySelectorOptions[testSelectorsKey],
+            ],
+          ),
+        ) as Record<
+          LocalFeatureTestSelector extends TestSelector
+            ? LocalFeatureTestSelector["selector"]["options"][keyof LocalFeatureTestSelector["selector"]["options"]]
+            : string,
+          string | RegExp
+        >,
+      };
+
+      capturedElement =
+        featureTestSelector.selector.method === "ByRole"
+          ? getByRoleQuerySelector({
+              ...sharedProps,
+              role:
+                Array.isArray(featureTestSelector.selector.role) || role
+                  ? role || ""
+                  : featureTestSelector.selector.role,
+            })
+          : getByTextQuerySelector({
+              ...sharedProps,
+              selectionMethod: featureTestSelector.selector.method,
+              text: featureTestSelector.selector.text,
+            });
+    } else if (
+      "isControlledElement" in featureTestSelector &&
+      featureTestSelector.isControlledElement
+    ) {
+      try {
+        capturedElement = getControlledElement({ element: containerElement });
+      } catch (error) {
+        if (queryMethod === "query") {
+          capturedElement = null;
+        }
+
+        throw error;
+      }
+    }
+
+    if (!capturedElement) {
+      throw new ElementError(
+        "No child HTML element available",
+        containerElement,
+      );
+    }
+
+    if (!("accessibleText" in featureTestSelector)) {
```

Comment:
This could be managed by the type system if we change the extends at the top of this function to `AccessibleLabelSelector & TestSelector`. Should user's be allowed to give a selector that doesn't have this already? Or is it expected that they should handle an error like this at runtime and recover from it?

Reply:
`TestSelector` isn't the only type that returns an element. It used to be, but now the `FeatureSelector` could have `isControlledElement`, and that also could follow you to an element.

You shouldn't be able to call this function if your type isn't valid, but the function itself doesn't know that. I tried using a ternary and `null` if it didn't have accessible text, but then I had to use `?.getAccessibleText` because TypeScript could never be sure the function existed. It's also a runtime check for `capturedElement` as it could be `null` if you use `"query"` as the `queryMethod`.

A few things about that:
1. It doesn't make sense that `isControlledElement` isn't another `selector` variant since it's related to element selection. `isControlledElement` used to be part of the feature itself rather than a `FeatureTestSelector` because you can't have a controlled element that isn't a child of another element. Also, does it make sense that a child element is a feature or that it's an associated element?
1. I should rename `TestSelector` to `ElementSelector` or something else. It was called `TestSelector` because it was a `TestingLibrarySelector` originally and `FeatureSelector` was a way to get features of the component. `TestSelector` should be more like the whole type rather than `FeatureTestSelector`. I'm gonna play around with names.
1. The `isControlledElement` check throws errors, and it should only throw when it's `get` rather than `query`. I have a `// TODO`, but I never yet fixed it in this branch.